controllers/component_dependency_update_controller.go

```diff
@@ -0,0 +1,293 @@
+/*
+Copyright 2021-2023 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	tektonapi "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1"
+	v12 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/utils/strings/slices"
+	"knative.dev/pkg/apis"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"strconv"
+	"time"
+
+	"github.com/go-logr/logr"
+	applicationapiv1alpha1 "github.com/redhat-appstudio/application-api/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	contextTimeout = 300 * time.Second
+	// PipelineRunTypeLabelName contains the type of the PipelineRunTypeLabelName.
+	PipelineRunTypeLabelName = "pipelines.appstudio.openshift.io/type"
+	// PipelineRunBuildType is the type denoting a build PipelineRun.
+	PipelineRunBuildType         = "build"
+	NudgeProcessedAnnotationName = "build.appstudio.openshift.io/component-nudge-processed"
+
+	// PipelineRunComponentLabelName is the label denoting the application.
+	PipelineRunComponentLabelName = "appstudio.openshift.io/component"
+	NudgeFinalizer                = "appstudio.openshift.io/build-nudge-finalizer"
+
+	PacEventTypeLabelName = "pipelinesascode.tekton.dev/event-type"
+	PacEventPush          = "push"
+	ComponentNudged       = "ComponentNudged"
+	ComponentNudgeFailed  = "ComponentNudgeFailed"
+	ImageUrlParam         = "IMAGE_URL"
+	ImageDigestParam      = "IMAGE_DIGEST"
+
+	TestFailureAnnotation  = "internal.testing.appstudio.openshift.io/unit-test-failure-count"
+	FailureCountAnnotation = "appstudio.openshift.io/build-nudge-failures"
+	MaxAttempts            = 3
+	FailureRetryTime       = time.Minute * 5 // We retry after 5 minutes on failure
```

Comment:
Not sure that we should wait 5*x minutes after a failure. It could be a network error as well.